### ISSUE TITLE
mossStore on-disk storage size

### ISSUE THREAD

**Issue by tcolgate** [blevesearch/bleve#553]
I am experimenting with indexing of log data with bleve. The data in question consists of:
* a timestamp, truncated to 1s intervals (indexed, not stored).
* a text field, using the simple analyzer (indexed, not stored).
* a set of KV pairs, using the keyworld analyzer (indexed, not stored).
* a filename, using the keyword analyzer (not indexed, stored)
* a string identifying the offset into the file (not indexed, stored).
* document IDs are numerical, incrementinng from 0
(I'll provide the mapping in a gist).

So basically I'm building an out of band index of a bunch of protobuf files containing the actual log data.
I've previously attempted storing the actual log data in the index (variously disabled bits of it), but in all cases, with all the storage backends, the sizes were prohibitively large. I'm keeping the serialized protobuf files anyway, so the above index works reasonably well for our needs.

In terms of total volumes of data, we have two installs, one doing approximately 50 msg/s, the other doing aroudn 400/s. Lines are usually a few hundres bytes, with around half a dozen kv pairs (or low cardinaity, around half a dozen unique labels, and maybe a two hundred values of those in total).
Data is stored in 15m shards at present,  Smaller install would be about 45k documents, largest install about 360k, per indexed shard.

I have testing various bleve storage options, only two coped well with the throughput, goleveldb and moss (with mossStore enabled).

moss shows more  CPU use , but much lower memory requirements. It seems to perform much better when searching an active shard (though I did notice a potential memory leak, I may be able to provide more details of that later). In terms of total throughput, moss appears to be the more performant of the two. Under benchmarking loads, goleveldb's compactions ruin the overall throughput (this is less evident under the real world loads mentioned above)

Unfortunately the on disk storage requirements of moss are considerably more than those for goleveldb. We see around 35MB for a 15 minute index of the smaller install using goleveldb, vs 150MB for mossStore (the raw, uncompressed protobuf data is around 15MB, though most labels are only stored once in the protobuf for a stream, not with every message).

![logspray](https://cloud.githubusercontent.com/assets/76019/24034910/10fd4280-0aeb-11e7-9196-890a124da275.png). 

The larger install seeing the same proportional increase.

The disk usage on the graph is calculated by walking the directory holding the data.

From the graphs, it's interesting that mossStore's compactions don't seem to manage to actuall compact the data at all (size after a compaction is shows the same upward trend), The only fields that are unique in each message are the ID, and the offsets (even some of the offsets are likely the same), many of the log lines do contain millisecond timestamps, but many do not.

I'd like to use mossStore, (the query performance of goleveldb seems considerably worse), but at present the additional storage size is too much of an overhead. Is this a known issue, or perhaps by design? Can I expect any significant improvement in the on disk overhead of moss?

**Comment by tcolgate** [comment]
Code for the mapping is here...
https://gist.github.com/anonymous/22801d6fb465dffcf73c3d0658dce079

**Comment by steveyen** [comment]
Hi @tcolgate,
Some thoughts...

One thought is that it seems you have time-series or log oriented data, where new data is continually added but old records aren't updated.  In this kind of situation, the compaction performed by mossStore (and by goleveldb, too; and the LSM family of stores) will end up reordering the various key-val data records (like merge-sorting the various append-only "segments" to construct a new segment).  But, there aren't any key-val records to de-duplicate (there aren't any updates that will obsolete old records) and/or there aren't any deleted records to throw away.  Hence, compaction ends up performing a reformatting/reordering of the key-val records to make queries faster, but not a reduction of data size.

The second thought is that mossStore, unlike goleveldb and some other storage libraries, unfortunately doesn't provided compression, yet.  On the "yet", adding compression to mossStore isn't a "quick hit turnaround" kind of feature though.  For example, (go)leveldb performs compression on a block-by-block basis, whereas mossStore, in contrast, doesn't even have a concept of blocks.  I've just created a tracking issue for compression over on the github.com/moss project... https://github.com/couchbase/moss/issues/20 ...at least as a place to gather and formulate design ideas.

cheers,
steve

**Comment by tcolgate** [comment]
Are sub-documents deduped? I originally moved the KV into the root document so that users would not need to search using `labels.thing`, however, we're not going to be using raw bleve string queries anymore, so moving them back to asub document is more viable.
  I can experiment with moving the KV pairs into a separate document type, and referencing those from the message objects. That should reduce the size of a row (will complicate querying somewhat).

**Comment by steveyen** [comment]
> Are sub-documents deduped?

Hi @tcolgate,
I think fields in nested parts of a document won't be deduplicated either based on your log-like use case, as overall you're just continually adding more and more new documents .  Unless I'm misunderstanding what you mean...

cheers,
steve

**Comment by tcolgate** [comment]
I'll try the separate document approach.

**Comment by tcolgate** [comment]
Since LSM seemed like the wrong tool for the job, I went back to the boltdb back end. This actually performed pretty well, but the on disk storage was still excessive. I didn't try the separate documents for KV set in the end, (since it complicates my query).
  Regrettably I've dropped bleve. Our query requirements are pretty simple, so for the moment I'm moving back to linear search of the data, and may do a simple reverse index in the longer term.
  FWIW, I guess a lot of people will attempt to use bleve for log data, I probabyl wont be the last. Having one storage back end that supports the approach well would probably be a good thing.
